File: Sources/Sentry/include/SentryDisplayLinkWrapper.m
#import "SentryDisplayLinkWrapper.h"

#if SENTRY_HAS_UIKIT

#    import <UIKit/UIKit.h>

@implementation SentryDisplayLinkWrapper {
    CADisplayLink *displayLink;
}

- (CFTimeInterval)timestamp
{
    return displayLink.timestamp;
}

- (CFTimeInterval)targetTimestamp API_AVAILABLE(ios(10.0), tvos(10.0))
{
    return displayLink.targetTimestamp;
}

- (void)linkWithTarget:(id)target
                    selector:(SEL)sel
    preferredFramesPerSecond:(NSInteger)preferredFramesPerSecond
{
    displayLink = [CADisplayLink displayLinkWithTarget:target selector:sel];
    if (preferredFramesPerSecond >= 0) {
        displayLink.preferredFramesPerSecond = preferredFramesPerSecond;
    }
    [displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
}

- (void)linkWithTarget:(id)target
                   selector:(SEL)sel
    preferredFrameRateRange:(CAFrameRateRange)preferredFrameRateRange API_AVAILABLE(ios(15.0))
{
    displayLink = [CADisplayLink displayLinkWithTarget:target selector:sel];
    displayLink.preferredFrameRateRange = preferredFrameRateRange;
    [displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
}

- (void)invalidate
{
    [displayLink invalidate];
    displayLink = nil;
}

- (BOOL)isRunning
{
    return displayLink != nil && !displayLink.isPaused;
}

@end

#endif // SENTRY_HAS_UIKIT

File: Sources/Swift/Integrations/Performance/SentryDisplayLinkWrapper.swift
// swiftlint:disable missing_docs
#if (os(iOS) || os(tvOS) || os(visionOS)) && !SENTRY_NO_UIKIT

import UIKit

/// A wrapper around DisplayLink for testability.
@_spi(Private) @objc public class SentryDisplayLinkWrapper: NSObject {
    
    var displayLink: CADisplayLink?

    @objc public var timestamp: CFTimeInterval {
        displayLink?.timestamp ?? 0
    }

    @objc public var targetTimestamp: CFTimeInterval {
        displayLink?.targetTimestamp ?? 0
    }

    @objc public func link(withTarget target: Any, selector sel: Selector) {
        displayLink = CADisplayLink(target: target, selector: sel)
        displayLink?.add(to: .main, forMode: .common)
    }
    
    @objc public func invalidate() {
        displayLink?.invalidate()
        displayLink = nil
    }
    
    @objc public func isRunning() -> Bool {
        !(displayLink?.isPaused ?? true)
    }
}

#if (os(iOS) || os(tvOS)) && !SENTRY_NO_UIKIT
extension SentryDisplayLinkWrapper: SentryReplayDisplayLinkWrapper {}
#endif

#endif
// swiftlint:enable missing_docs
